Source/cryptalgo/SecureSocketPort.cpp

```diff
@@ -131,36 +131,46 @@ SecureSocketPort::Handler::~Handler() {
 }
 
 uint32_t SecureSocketPort::Handler::Initialize() {
-    uint32_t success = Core::ERROR_NONE;
-
-    _context = SSL_CTX_new(TLS_method());
-
-    _ssl = SSL_new(static_cast<SSL_CTX*>(_context));
-    SSL_set_fd(static_cast<SSL*>(_ssl), static_cast<Core::IResource&>(*this).Descriptor());
-    SSL_CTX_set_options(static_cast<SSL_CTX*>(_context), SSL_OP_ALL | SSL_OP_NO_SSLv2 | SSL_OP_NO_SSLv3);
-
-    // Trust the same certificates as any other application
-    if (SSL_CTX_set_default_verify_paths(static_cast<SSL_CTX*>(_context)) == 1) {
-        success = Core::SocketPort::Initialize();
-    } else {
-        TRACE_L1("OpenSSL failed to load certificate store");
-        success = Core::ERROR_GENERAL;
+    uint32_t success = Core::ERROR_GENERAL;
+
+    if ((_context = SSL_CTX_new(TLS_method())) != nullptr) {
+        if (   // Returns bit-mask after adding options
+               ((SSL_CTX_set_options(static_cast<SSL_CTX*>(_context), SSL_OP_ALL | SSL_OP_NO_SSLv2 | SSL_OP_NO_SSLv3) & (SSL_OP_ALL | SSL_OP_NO_SSLv2 | SSL_OP_NO_SSLv3))  == (SSL_OP_ALL | SSL_OP_NO_SSLv2 | SSL_OP_NO_SSLv3))
+            && // Trust the same certificates as any other application
+               (SSL_CTX_set_default_verify_paths(static_cast<SSL_CTX*>(_context)) == 1)
+            && ((_ssl = SSL_new(static_cast<SSL_CTX*>(_context))) != nullptr)
+            && (SSL_set_fd(static_cast<SSL*>(_ssl), static_cast<Core::IResource&>(*this).Descriptor()) == 1)
+        ) {
+            success = Core::SocketPort::Initialize();
+        } else {
+            TRACE_L1("OpenSSL failed to set protocol level or load certificate store");
+            success = Core::ERROR_GENERAL;
+        }
     }
 
     return success;
 }
 
 int32_t SecureSocketPort::Handler::Read(uint8_t buffer[], const uint16_t length) const {
-    int32_t result = SSL_read(static_cast<SSL*>(_ssl), buffer, length);
+    int32_t result = 0;
 
-    if (_handShaking != CONNECTED) {
-        const_cast<Handler&>(*this).Update();
+    if (_handShaking == CONNECTED) {
+        int value = SSL_read(static_cast<SSL*>(_ssl), buffer, length);
+        result = (value > 0 ? value : 0);
     }
+
     return (result);
 }
 
 int32_t SecureSocketPort::Handler::Write(const uint8_t buffer[], const uint16_t length) {
-    return (SSL_write(static_cast<SSL*>(_ssl), buffer, length));
+    int32_t result = 0;
+
+    if (_handShaking == CONNECTED) {
```

Comment:
This was/is a design flaw, if someone would write already while there is no "OPEN"/"CONNECTED state, it might end up in a busy-loop. Suggest an ASSERT( _handShaking == CONNECTED); Programitically the user of the socket has to wait till the StateChange -> Open has ocured before the Write can be used!!